### PR TITLE
Added Top K frequent elements under heap section.

### PR DESCRIPTION
**Pull request by tanyarajhans**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

### Description

Added Top K frequent elements under heap section of C-Plus-Plus and updated the Readme as well.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

https://github.com/TesseractCoding/NeoAlgo/issues/2252

### CODE REVIEW

**Comment by tanyarajhans** [comment]
@HarshCasper please review.

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@tanyarajhans Remove whitespaces from code and add problem link. Format like this

```c++
/*
Top K frequent Elements
======================================
Given a non-empty array of integers, return the k most frequent elements.

Steps:

1.We will build a map and store the frequency of every element in the array.
2.Now we will iterate over the map and push all the elements along with their frequencies in a maxHeap as a pair.
3.Since when pushed as a pair, elements are sorted in descending order(due to maxHeap) based on the first element of the
pair, therefore, we will make the pair as {frequency of the element, element}.
4.After we have pushed all the elements into the priority queue, we will run a while loop till k is greater than 0 and the priority queue is not empty and pop the elements and print them.

You may assume k is always valid, 1 ≤ k ≤ number of unique elements.

Time Complexity: O(nlogn)
Space Complexity: O(n)
*/

#include <bits/stdc++.h>
using namespace std;

void topKFrequent(vector<int>& nums, int k) {
 map<int, int>m;
 priority_queue<pair<int, int>> pq;

 for (int i : nums) {
 m[i]++;
 }

 for (auto i : m) {
 pq.push({i.second, i.first});
 }

 while (k-- && !pq.empty()) {
 cout << pq.top().second << " ";
 pq.pop();
 }
}

// Driver code
int main() {

 // Take input
 int n;
 cin >> n;
 vector<int> nums(n);
 for (int i = 0; i < n; i++) {
 cin >> nums[i];
 }
 int k;
 cin >> k;
 topKFrequent(nums, k);
}

/*
Input:
n = 8
nums = [1,1,7,4,2,2,2,3]
k = 2
Output: [2,1]

Explanation:
The top 2 frequent elements are 2 and 1 as 2 occurs 3 times and 1 occurs 2 times.
*/
```

File: C-Plus-Plus/heap/Top_K_frequent_elements.cpp [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+1.We will build a map and store frequency of every element in the array.
+2.Now we will iterate over the map and push all the elements along with their frequencies in a maxHeap as a pair. 
+3.Since when pushed as a pair, elements are sorted in descending order(due to maxHeap) based on the first element of the 
+pair, therefore, we will make the pair as {frequency of the element, element}.
+4.After we have pushed all the elements into the priority queue, we will run a while loop till k is greater than 0 and priority queue is not empty and pop the elements and print them.
+
+You may assume k is always valid, 1 ≤ k ≤ number of unique elements.
+
+Time Complexity: O(nlogn)
+Space Complexity: O(n)
```
Comment (HarshCasper) [inline C-Plus-Plus/heap/Top_K_frequent_elements.cpp]: Push Complexity to the end of the Code

File: C-Plus-Plus/heap/Top_K_frequent_elements.cpp [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+======================================
+Given a non-empty array of integers, return the k most frequent elements.
+
+
+Steps:
+1.We will build a map and store frequency of every element in the array.
+2.Now we will iterate over the map and push all the elements along with their frequencies in a maxHeap as a pair. 
+3.Since when pushed as a pair, elements are sorted in descending order(due to maxHeap) based on the first element of the 
+pair, therefore, we will make the pair as {frequency of the element, element}.
+4.After we have pushed all the elements into the priority queue, we will run a while loop till k is greater than 0 and priority queue is not empty and pop the elements and print them.
```
Comment (HarshCasper) [inline C-Plus-Plus/heap/Top_K_frequent_elements.cpp]: Modularize this line and break it up for better understanding

File: C-Plus-Plus/heap/Top_K_frequent_elements.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
+/*
+Top K frequent Elements
+======================================
+Given a non-empty array of integers, return the k most frequent elements.
+
```
Comment (HarshCasper) [inline C-Plus-Plus/heap/Top_K_frequent_elements.cpp:5]: Add the Leetcode question link. It was one of my Interview Questions as well

File: C-Plus-Plus/heap/Top_K_frequent_elements.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        m[i]++;
+    }
+        
+    for(auto i : m){
+        pq.push({i.second, i.first});
+    }
+
+    while(k-- && !pq.empty()){
+        cout<<pq.top().second<<" ";
+        pq.pop();
```
Comment (HarshCasper) [inline C-Plus-Plus/heap/Top_K_frequent_elements.cpp]: Document it

**Review by HarshCasper — CHANGES_REQUESTED** [review]
Rest looks good

File: C-Plus-Plus/heap/Top_K_frequent_elements.cpp (hunk 4 above)
Comment (tanyarajhans) [reply C-Plus-Plus/heap/Top_K_frequent_elements.cpp]: I didn't understand? I have explained this step earlier.

**Comment by tanyarajhans** [comment]
@iamrajiv @HarshCasper please review.

**Review by HarshCasper — APPROVED** [review]

**Review by ankitaggarwal23 — COMMENTED** [review]
Code Plagiarized from [here](https://leetcode.com/problems/top-k-frequent-elements/discuss/81631/3-ways-to-solve-this-problem/363016/).

**Comment by tanyarajhans** [comment]
> Code Plagiarized from [here](https://leetcode.com/problems/top-k-frequent-elements/discuss/81631/3-ways-to-solve-this-problem/363016/).

This is my own code which I wrote and submitted on LeetCode. https://leetcode.com/submissions/detail/465190983/
You should be aware that I can't produce a new algorithm and study from internet resources and learn but write my own code. I don't understand why you would do this but go ahead! Also just because someone famous wrote an article on LeetCode using the same idea, doesn't mean my brain can't comprehend the same idea on my own. I have solved more than 1200+ problems and one of my articles trended at #1 on LeetCode for a week!

**Comment by HarshCasper** [comment]
> > Code Plagiarized from [here](https://leetcode.com/problems/top-k-frequent-elements/discuss/81631/3-ways-to-solve-this-problem/363016/).
> 
> This is my own code which I wrote and submitted on LeetCode. https://leetcode.com/submissions/detail/465190983/
> You should be aware that I can't produce a new algorithm and study from internet resources and learn but write my own code. I don't understand why you would do this but go ahead! Also just because someone famous wrote an article on LeetCode using the same idea, doesn't mean my brain can't comprehend the same idea on my own. I have solved more than 1200+ problems and one of my articles trended at #1 on LeetCode for a week!

How does solving 1200+ problems on LeetCode and having an article trending on LeetCode matter anything to us? Both of them are vanity metrics, which would look good on a portfolio, but won't matter much here. 

Let's go back to the basics: You made 3 PRs, of which two are clear-cut cases of plagiarism and one has its primary logic taken up and modified. The least we are doing is to revert your PRs back, so that some doesn't trace us up and accuse us of having copied code from around the Internet collected over here. 

We are not asking you to come up with your own algorithm. Almost every contributor is contributing by studying existing resources, while we have some who love to jump over the bandwagon and push whatever they desire. 

The someone "famous" you are referring to made the post back in 2018. I don't know how come you are defending this fact. 

The whole argument reeks of arrogance at the best and total disregard for standard rules and conventions. It was unfortunate that our guard was lower when we were taking initial contributions up and now have to deal with a lot of mess due to plagiarism. 

Thanks for your contribution and best of luck with your Open-Source Journey further! And happy LeetCoding. Let me know once you complete all the 1800 LeetCode Questions. Good Luck!

**Pull request merged**